larpandoracontent/LArThreeDReco/LArCosmicRay/TwoViewCosmicRayRemovalTool.cc

```diff
@@ -0,0 +1,560 @@
+
+            modifiedClusters.insert(pDeltaRayCluster);
+
+            this->SplitDeltaRayCluster(element, hitType, deltaRayHits, deltaRayRemnantHits);
+        }
+    }
+
+    changesMade = modifiedClusters.size();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::PassElementChecks(const MatrixType::Element &element, const HitType hitType) const
+{
+    // ATTN: Avoid endpoints, topological michel reconstruction is very ambiguous
+    if (this->IsMuonEndpoint(element, hitType))
+        return false;
+
+    const Cluster *pMuonCluster(nullptr), *const pDeltaRayCluster(m_pParentAlgorithm->GetCluster(element, hitType));
+    
+    if (m_pParentAlgorithm->GetMuonCluster(element.GetOverlapResult().GetCommonMuonPfoList(), hitType, pMuonCluster) != STATUS_CODE_SUCCESS)
+        return false;
+
+    const float separation(LArClusterHelper::GetClosestDistance(pDeltaRayCluster, pMuonCluster));
+    
+    if (separation > m_minSeparation)
+        return false;
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::IsMuonEndpoint(const MatrixType::Element &element, const HitType hitType) const
+{
+    for (const HitType &otherHitType : m_pParentAlgorithm->GetHitTypeVector())
+    {
+        if (otherHitType == hitType)
+            continue;
+
+        const Cluster *pMuonCluster(nullptr), *const pDeltaRayCluster(m_pParentAlgorithm->GetCluster(element, hitType));
+
+        if (m_pParentAlgorithm->GetMuonCluster(element.GetOverlapResult().GetCommonMuonPfoList(), hitType, pMuonCluster) != STATUS_CODE_SUCCESS)
+            return false;
+
+        float xMinDR(+std::numeric_limits<float>::max()), xMaxDR(-std::numeric_limits<float>::max());
+        pDeltaRayCluster->GetClusterSpanX(xMinDR, xMaxDR);
+
+        float xMinCR(-std::numeric_limits<float>::max()), xMaxCR(+std::numeric_limits<float>::max());
+        pMuonCluster->GetClusterSpanX(xMinCR, xMaxCR);
+
+        if ((xMinDR < xMinCR) || (xMaxDR > xMaxCR))
+            return false;
+	
+        float zMinDR(+std::numeric_limits<float>::max()), zMaxDR(-std::numeric_limits<float>::max());
+        pDeltaRayCluster->GetClusterSpanZ(xMinDR, xMaxDR, zMinDR, zMaxDR);
+
+        float zMinCR(-std::numeric_limits<float>::max()), zMaxCR(+std::numeric_limits<float>::max());
+        pMuonCluster->GetClusterSpanZ(xMinCR, xMaxCR, zMinCR, zMaxCR);
+
+        if ((zMinDR < zMinCR) || (zMaxDR > zMaxCR))
+            return false;
+    }
+
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewCosmicRayRemovalTool::IsContaminated(const MatrixType::Element &element, const HitType hitType) const
```

Comment:
Also essentially the same as the CosmicRayRemovalTool version. It appears to be just the element type that may be different, I'd need to check the relationship between the different types here, but I suspect this could similarly be consolidated into a base class.